p2p/http/libp2phttp.go

```diff
@@ -0,0 +1,744 @@
+			// resolve the host
+			ipaddr, err := net.ResolveIPAddr("ip", parsedAddr.host)
+			if err != nil {
+				return err
+			}
+
+			host := ipaddr.String()
+			l, err := net.Listen("tcp", host+":"+parsedAddr.port)
+			if err != nil {
+				return err
+			}
+			h.httpTransport.listeners = append(h.httpTransport.listeners, l)
+
+			// get resolved port
+			_, port, err := net.SplitHostPort(l.Addr().String())
+			if err != nil {
+				return err
+			}
+
+			var listenAddr ma.Multiaddr
+			if parsedAddr.useHTTPS && parsedAddr.sni != "" && parsedAddr.sni != host {
+				listenAddr = ma.StringCast(fmt.Sprintf("/ip4/%s/tcp/%s/tls/sni/%s/http", host, port, parsedAddr.sni))
+			} else {
+				scheme := "http"
+				if parsedAddr.useHTTPS {
+					scheme = "https"
+				}
+				listenAddr = ma.StringCast(fmt.Sprintf("/ip4/%s/tcp/%s/%s", host, port, scheme))
+
+			}
+
+			if parsedAddr.useHTTPS {
+				go func() {
+					srv := http.Server{
+						Handler:   &h.ServeMux,
+						TLSConfig: h.TLSConfig,
+					}
+					errCh <- srv.ServeTLS(l, "", "")
+				}()
+				h.httpTransport.listenAddrs = append(h.httpTransport.listenAddrs, listenAddr)
+			} else if h.ServeInsecureHTTP {
+				go func() {
+					errCh <- http.Serve(l, &h.ServeMux)
+				}()
+				h.httpTransport.listenAddrs = append(h.httpTransport.listenAddrs, listenAddr)
+			} else {
+				// We are not serving insecure HTTP
+				log.Warnf("Not serving insecure HTTP on %s. Prefer an HTTPS endpoint.", listenAddr)
+			}
+		}
+		return nil
+	}()
+	if err != nil {
+		closeAllListeners()
+		return err
+	}
+
+	close(h.httpTransport.waitingForListeners)
+	closedWaitingForListeners = true
+
+	if len(h.httpTransport.listeners) == 0 || len(h.httpTransport.listenAddrs) == 0 {
+		closeAllListeners()
+		return ErrNoListeners
+	}
+
+	expectedErrCount := len(h.httpTransport.listeners)
+	select {
+	case <-h.httpTransport.closeListeners:
+	case err = <-errCh:
+		expectedErrCount--
+	}
+
+	// Close all listeners
+	closeAllListeners()
+	for i := 0; i < expectedErrCount; i++ {
+		<-errCh
+	}
+	close(errCh)
+
+	return err
+}
+
+func (h *HTTPHost) Close() error {
+	h.createHTTPTransport.Do(func() {
+		h.httpTransport = newHTTPTransport()
+	})
+	close(h.httpTransport.closeListeners)
+	return nil
+}
+
+// SetHTTPHandler sets the HTTP handler for a given protocol. Automatically
+// manages the .well-known/libp2p mapping.
+// http.StripPrefix is called on the handler, so the handler will be unaware of
+// it's prefix path.
+func (h *HTTPHost) SetHTTPHandler(p protocol.ID, handler http.Handler) {
+	h.SetHTTPHandlerAtPath(p, string(p), handler)
+}
+
+// SetHTTPHandlerAtPath sets the HTTP handler for a given protocol using the
+// given path. Automatically manages the .well-known/libp2p mapping.
+// http.StripPrefix is called on the handler, so the handler will be unaware of
+// it's prefix path.
+func (h *HTTPHost) SetHTTPHandlerAtPath(p protocol.ID, path string, handler http.Handler) {
+	if path[len(path)-1] != '/' {
```

Comment:
If an empty `path` is passed in, this will panic. Caller will need to check for empty `path`. Maybe safer to handle the `len(path) == 0` case?

Reply:
Also typo on it’s